Make conda and python builds faster

arthurp opened the pull request.

witchel commented on libtsuba/src/S3Storage.cpp:
```diff
@@ -0,0 +1,129 @@
+  auto [bucket, object] = S3SplitUri(std::string(uri));
+  if (bucket.empty() || object.empty()) {
+    return ErrorCode::InvalidArgument;
+  }
```
I wonder if it is worth having a function similar to CleanUri that all S3 functions call in the start.  Maybe
auto [res, bucket, object] = S3CleanUri(uri);
if(!res) { return Res.error() }

tylershunt replied:
that's a good idea.

witchel commented on libtsuba/src/file.cpp:
```diff
@@ -11,23 +11,16 @@
 #include <fstream>
 #include <iostream>
 #include <cassert>
-#include <boost/filesystem.hpp>
```
Oh yeah!  Bye boost/filesystem.ccp from file.

tylershunt commented on libtsuba/include/tsuba/Errors.h:
```diff
@@ -16,6 +16,7 @@ enum class ErrorCode {
   Exists          = 5,
   OutOfMemory     = 6,
   TODO            = 7,
+  StorageError    = 8,
```
Yes, i think all of the current `StorageError`s are s3 errors anyway. So that seems reasonable. Yeah all of the local file system errors are propagated by `ResultErrno`s.

arthurp committed:
Add caching for katana_python.py requirements checks.

This parallels CMakeCache.txt. But is dumber. But still makes my
local builds a bit faster.

arthurp committed:
Use cmake CLI to build and install in conda build.

For reasons I do not understand and cannot reproduce, make install
was causing a complete recompile in he conda build. My only guess
is that some option in the conda environment causes a -D to change
in the build flags during installation forcing a recompile. But,
honestly that seems super unlikely.

arthurp committed:
Make ccache ignore the compiler mtime (and cwd) to hit more.
